File: erp-mobile-app/src/lib/counterDataIsolation.ts
/** Client-side row scoping when counter worker is active under Admin JWT. */



import {

  resolveModuleListBranchScope,

  type ListBranchScope,

} from './listBranchScope';



export type { ListBranchScope };



export function resolveCounterListBranchScope(

  selectedBranchId: string | null | undefined,

  accessibleBranchIds: string[],

  isAdminOrOwner: boolean,

  isolateWorkerData?: boolean,

): ListBranchScope {

  return resolveModuleListBranchScope(

    selectedBranchId,

    accessibleBranchIds,

    isAdminOrOwner,

    isolateWorkerData,

  );

}



export function shouldIsolateCounterWorkerData(role: string): boolean {

  const r = (role || '').toLowerCase();

  return r === 'worker' || r === 'salesman';

}



function normalizeId(value: unknown): string | null {

  if (typeof value !== 'string') return null;

  const trimmed = value.trim();

  return trimmed ? trimmed.toLowerCase() : null;

}



function collectRowCreatorIds(row: Record<string, unknown>): Set<string> {

  const ids = new Set<string>();

  for (const key of ['created_by_id', 'created_by', 'user_id', 'paid_to_user_id', 'salesman_id'] as const) {

    const normalized = normalizeId(row[key]);

    if (normalized) ids.add(normalized);

  }

  return ids;

}



function collectWorkerIdentityIds(authUserId: string, profileId?: string | null): Set<string> {

  const ids = new Set<string>();

  const auth = normalizeId(authUserId);

  if (auth) ids.add(auth);

  const profile = normalizeId(profileId);

  if (profile) ids.add(profile);

  return ids;

}



export function rowBelongsToCounterWorker(

  row: Record<string, unknown>,

  authUserId: string,

  profileId?: string | null,

): boolean {

  const workerIds = collectWorkerIdentityIds(authUserId, profileId);

  if (workerIds.size === 0) return false;

  const rowCreatorIds = collectRowCreatorIds(row);

  for (const id of rowCreatorIds) {

    if (workerIds.has(id)) return true;

  }

  return false;

}

/** Rental row visible to worker: created_by or salesman_id matches identity. */
export function rowBelongsToRentalWorker(
  row: Record<string, unknown>,
  authUserId: string,
  profileId?: string | null,
): boolean {
  if (rowBelongsToCounterWorker(row, authUserId, profileId)) return true;
  const workerIds = collectWorkerIdentityIds(authUserId, profileId);
  const salesman = normalizeId(row.salesman_id ?? row.salesmanId);
  return salesman != null && workerIds.has(salesman);
}

File: erp-flutter-v2/lib/core/permissions/counter_data_isolation.dart
import '../../features/auth/providers/permission_state.dart';

/// Mirrors erp-mobile-app/src/lib/counterDataIsolation.ts
bool shouldIsolateCounterWorkerData(String role) {
  final r = role.toLowerCase();
  return r == 'worker' || r == 'salesman';
}

String? _normalizeId(dynamic value) {
  if (value is! String) return null;
  final trimmed = value.trim();
  return trimmed.isEmpty ? null : trimmed.toLowerCase();
}

Set<String> _workerIdentityIds(String authUserId, String? profileId) {
  final ids = <String>{};
  final auth = _normalizeId(authUserId);
  if (auth != null) ids.add(auth);
  final profile = _normalizeId(profileId);
  if (profile != null) ids.add(profile);
  return ids;
}

Set<String> _rowCreatorIds(Map<String, dynamic> row) {
  final ids = <String>{};
  for (final key in [
    'created_by_id',
    'created_by',
    'user_id',
    'paid_to_user_id',
    'salesman_id',
  ]) {
    final normalized = _normalizeId(row[key]);
    if (normalized != null) ids.add(normalized);
  }
  return ids;
}

bool rowBelongsToCounterWorker(
  Map<String, dynamic> row,
  String authUserId,
  String? profileId,
) {
  final workerIds = _workerIdentityIds(authUserId, profileId);
  if (workerIds.isEmpty) return false;
  final creators = _rowCreatorIds(row);
  for (final id in creators) {
    if (workerIds.contains(id)) return true;
  }
  return false;
}

bool shouldIsolateSalesList(PermissionState perms, String role) {
  if (perms.isAdminOrOwner) return false;
  if (perms.hasPermission('sales.view_company') ||
      perms.hasPermission('sales.view_branch')) {
    return false;
  }
  return perms.hasPermission('sales.view_own') ||
      shouldIsolateCounterWorkerData(role);
}

bool shouldIsolatePurchaseList(PermissionState perms, String role) {
  if (perms.isAdminOrOwner) return false;
  if (perms.hasPermission('purchase.view_company') ||
      perms.hasPermission('purchase.view_branch')) {
    return false;
  }
  return perms.hasPermission('purchase.view_own') ||
      shouldIsolateCounterWorkerData(role);
}
